dev/src/main/java/org/eclipse/codewind/intellij/ui/toolwindow/UpdateHandler.java

```diff
@@ -0,0 +1,163 @@
+                        listener.update();
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public void removeConnection(List<CodewindApplication> apps) {
+        synchronized(appListeners) {
+            for (CodewindApplication application : apps) {
+                Set<AppKey> appKeys = appListeners.keySet();
+                for (AppKey key : appKeys) {
+                    if (application.connection.getConid().equals(key.connectionId) &&
+                            application.projectID.equals(key.projectId)) {
+                        AppUpdateListener listener = appListeners.get(key);
+                        if (listener != null) {
+                            listener.remove();
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public void removeApplication(CodewindApplication application) {
+        synchronized(appListeners) {
+            Set<AppKey> appKeys = appListeners.keySet();
+            for (AppKey key : appKeys) {
+                if (application.connection.getConid().equals(key.connectionId) &&
+                        application.projectID.equals(key.projectId)) {
+                    AppUpdateListener listener = appListeners.get(key);
+                    if (listener != null) {
+                        listener.remove();
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Any Tool Window Content can be added as a listener for any application updates
+     * To identify the correct window, the following arguments are required to differentiate
+     *
+     * @param content
+     * @param connectionId
+     * @param projectID
+     * @param listener
+     */
+    public void addAppUpdateListener(Content content, String connectionId, String projectID, AppUpdateListener listener) {
+        synchronized(appListeners) {
+            appListeners.put(new AppKey(content, connectionId, projectID), listener);
+        }
+    }
+
+    public void removeAppUpdateListener(Content content, String connectionId, String projectID) {
+        synchronized(appListeners) {
+            Set<AppKey> appKeys = appListeners.keySet();
+            List<AppKey> keysToRemove = new ArrayList<>();
+            for (AppKey key : appKeys) {
+                AppUpdateListener listener = appListeners.get(key);
+                if (content == key.content && key.connectionId.equals(connectionId) && key.projectId.equals(projectID)) {
+                    if (listener != null) {
+                        listener.remove();
+                        keysToRemove.add(key);
+                    }
+                }
+            }
+            for (AppKey key : keysToRemove) {
+                appListeners.remove(key);
+            }
+        }
+    }
+
+    public interface AppUpdateListener {
+        public void update();
+        public void remove();
+    }
+
+    private class AppKey {
```

Comment:
`AppKey` can be a static nested class instead of an inner class. Change this to `private static class AppKey { ...`